### PR TITLE
Upgrade to Python 3.7.1

### PR DESCRIPTION
**Pull request by aseering**
https://github.com/iodide-project/pyodide/pull/594 tries to upgrade from Python 3.7.0 to 3.7.6.  The direct path results in a bunch of tests failing for what I believe are multiple different reasons.  

I'm not familiar enough with this codebase to feel comfortable tackling all of those issues in one shot myself.  So I propose to take a more incremental approach.

Several of the patch versions between 3.7.0 and 3.7.6 appear to introduce distinct test failures.  So, one at a time?

This PR upgrades from Python 3.7.0 to 3.7.1.  It also handles two test failures that are introduced by this upgrade.  If y'all would be willing, I would appreciate a review on whether this is the correct approach.

### CODE REVIEW

**Comment by aseering** [comment]
This PR has two commits:
* Upgrade from Python 3.7.0 to 3.7.1.  This is taken with only the obvious trivial modifications from https://github.com/iodide-project/pyodide/pull/594
* Work around failures in introduced tests.

File: cpython/patches/testing.patch [hunk 1]
```diff
@@ -82,3 +82,30 @@ index 1fc4de11e1..e6c91707ae 100644
...
+     def test_decompress_stdin_stdout(self):
++        raise unittest.SkipTest('no subprocess')
+         with io.BytesIO() as bytes_io:
+             with gzip.GzipFile(fileobj=bytes_io, mode='wb') as gzip_file:
+                 gzip_file.write(self.data)
+@@ -727,6 +729,7 @@
+ 
+     @create_and_remove_directory(TEMPDIR)
+     def test_compress_stdin_outfile(self):
++        raise unittest.SkipTest('no subprocess')
```
Comment (aseering) [inline cpython/patches/testing.patch]: Test failure:

```
======================================================================
ERROR: test_compress_stdin_outfile (test.test_gzip.TestCommandLine)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/lib/python3.7/test/test_gzip.py", line 683, in wrapper
    return function(*args, **kwargs)
  File "/lib/python3.7/test/test_gzip.py", line 731, in test_compress_stdin_outfile
    with Popen(args, stdin=PIPE, stdout=PIPE, stderr=PIPE) as proc:
  File "/lib/python3.7/subprocess.py", line 769, in __init__
    restore_signals, start_new_session)
  File "/lib/python3.7/subprocess.py", line 1447, in _execute_child
    restore_signals, start_new_session, preexec_fn)
BlockingIOError: [Errno 11] Resource temporarily unavailable

======================================================================
ERROR: test_decompress_stdin_stdout (test.test_gzip.TestCommandLine)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/lib/python3.7/test/test_gzip.py", line 699, in test_decompress_stdin_stdout
    with Popen(args, stdin=PIPE, stdout=PIPE, stderr=PIPE) as proc:
  File "/lib/python3.7/subprocess.py", line 769, in __init__
    restore_signals, start_new_session)
  File "/lib/python3.7/subprocess.py", line 1447, in _execute_child
    restore_signals, start_new_session, preexec_fn)
BlockingIOError: [Errno 11] Resource temporarily unavailable

----------------------------------------------------------------------
```
https://pastebin.com/TifZ3vEh for the full output -- this failure appears to cause a secondary failure which might also be interesting?

These two tests attempt to `Popen()` the current Python process and do stuff with it.  Several other tests in this suite are already skipped due to trying to do this, though the skip is implicit / buried in a helper function.

Several tests in other files successfully `Popen()` Python.  So it looks to me like it is possible to launch a subprocess in pyodide?  Is the difference that this case is opening stdin as a pipe?  What's up with subprocesses and pipes?; is anyone thinking about that currently?

**Comment by aseering** [comment]
This PR is currently failing a test in `test-chrome`:

```
_____________________________ test_brentq[chrome] ______________________________

selenium_standalone = <conftest.ChromeWrapper object at 0x7fb891b96be0>

    def test_brentq(selenium_standalone):
        selenium_standalone.load_package("scipy")
>       selenium_standalone.run("from scipy.optimize import brentq")
(...)
            raise exception_class(message, screen, stacktrace, alert_text)
>       raise exception_class(message, screen, stacktrace)
E       selenium.common.exceptions.WebDriverException: Message: unknown error: 
E         (Session info: headless chrome=70.0.3538.67)
E         (Driver info: chromedriver=2.41.578700 (2f1ed5f9343c13f73144538f15c00b370eda6706),platform=Linux 4.15.0-1052-aws x86_64)

/usr/local/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py:242: WebDriverException
```

However, the same failure currently happens on master, as can be seen here:

https://github.com/aseering/pyodide/pull/7

So I believe this failure is not an issue on the current branch.

**Comment by aseering** [comment]
FYI Python 3.7.2 has some additional failing tests that look at a quick first glance like they are probably failing for the same reason.  So it'd be good to confirm that this PR's approach of skipping those tests is as intended for now.

File: cpython/patches/testing.patch [hunk 2]
```diff
@@ -82,3 +82,30 @@ index 1fc4de11e1..e6c91707ae 100644
...
+ from test.support.script_helper import assert_python_ok
++import unittest
+ 
+ gzip = support.import_module('gzip')
+ 
+@@ -691,6 +692,7 @@
+     data = b'This is a simple test with gzip'
+ 
+     def test_decompress_stdin_stdout(self):
++        raise unittest.SkipTest('no subprocess')
```
Comment (rth) [inline cpython/patches/testing.patch]: Are you sure about this one? I wouldn't expect the test below to use subprocess..

Comment (rth) [reply cpython/patches/testing.patch]: Never mind I see you also included the original traceback in your comment https://github.com/iodide-project/pyodide/pull/619#discussion_r389400332

File: cpython/patches/testing.patch (hunk 1 above)
Comment (rth) [reply cpython/patches/testing.patch]: subprocessing should not supported as far as I know and they should be skipped via the flags in https://github.com/iodide-project/pyodide/blob/master/test/python_tests.txt file.

**Comment by rth** [comment]
Currently the list of CPython tests that are run is determined by [test/python_tests.txt](https://github.com/iodide-project/pyodide/blob/master/test/python_tests.txt). Could you please,
 - run `test/make_test_list.py` and examine the diff between the old and the new `test/python_tests.txt`
 - add the added tests in 3.7.1 (if any) to `test/python_tests.txt` and skip them if necessary with the `subprocess` tag

This allows to avoiding patching cpython tests that we can skip. 

Otherwise this PR LGTM, thanks!

**Comment by rth** [comment]
On a second though using flags in `test/python_tests.txt` would have required us to skip the whole `test_gzip` set of tests, which we don't want to do. So the current more granular approach with patching is better. We just need to make sure that no new tests were added using the procedure in my previous comment.

**Comment by aseering** [comment]
Thanks @rth for the review!

Regarding the procedure above, I ran the following command; is this what you mean?

```
$ python test/make_test_list.py 
$ git diff
$ 
```

(`python test/make_test_list.py` didn't produce any output, and it doesn't appear to have modified any files in the repo.)

**Comment by rth** [comment]
> (python test/make_test_list.py didn't produce any output, and it doesn't appear to have modified any files in the repo.)

Great! The `python_tests.txt` should have been created in the current directory I think. You should then be able to run,
```
diff python_tests.txt test/python_tests.txt
```

**Comment by aseering** [comment]
Ah, I see, there it is!  Thanks for the explanation.

Interestingly, that output file is empty.  I took a look at the script; I don't quite understand how its path-discovery logic works?  I was able to get the script to produce output by patching it as follows:

```
diff --git a/test/make_test_list.py b/test/make_test_list.py
index f735362..d347e0a 100644
--- a/test/make_test_list.py
+++ b/test/make_test_list.py
@@ -6,7 +6,7 @@ import os
 from pathlib import Path
 
 
-TEST_DIR = (Path(__file__).parent
+TEST_DIR = (Path(__file__).absolute().parent.parent
             / "cpython/build/3.7.1/host/lib/python3.7/test")
 
 
```

I then actually get quite a large diff:  https://pastebin.com/AkkF2PH0

It looks like most of the diff is that `test/python_tests.txt` has reason codes for disabled suites and `python_tests.txt` does not.  But I do see a few new tests in the list:

```
+test_asyncio.test_buffered_proto
+test_asyncio.test_context
+test_asyncio.test_runners
+test_asyncio.test_server
```

Is the correct procedure to add these to the file and see if they pass?  I'll give that a try.

**Comment by rth** [comment]
> Is the correct procedure to add these to the file and see if they pass? I'll give that a try.

Yes, exactly. Those will likely fail and can be skipped with the `async` tag.

**Comment by aseering** [comment]
Passing tests!  Would someone be willing to go ahead and merge?

**Pull request merged**